### PR TITLE
PN-24: Better seeding of similar patients by matching the full HPO hierarchies of the reference patient's symptoms to the index.

### PR DESCRIPTION
**Pull request by veronikaslc**
Changed generate search query function to search by all patient phenotype_ancestors list.

### CODE REVIEW

**Comment by veronikaslc** [comment]
@antonkatz , please review

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java [hunk 1]
```diff
@@ -151,16 +155,26 @@ public int compare(PatientSimilarityView o1, PatientSimilarityView o2)
...
+
         SolrQuery query = new SolrQuery();
         StringBuilder q = new StringBuilder();
-        // FIXME This is a very basic implementation, to be revisited
-        for (Feature phenotype : referencePatient.getFeatures()) {
-            if (StringUtils.isNotBlank(phenotype.getId())) {
-                q.append(phenotype.getType() + ":" + ClientUtils.escapeQueryChars(phenotype.getId()) + " ");
-            }
+        Iterator<String> iterator = ancestorsList.iterator();
+        q.append("(phenotype_ancestors" + ":" + ClientUtils.escapeQueryChars(iterator.next()));
```
Comment (antonkatz) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: I don't think you need the plus in ```"(phenotype_ancestors" + ":"```

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java [hunk 2]
```diff
@@ -151,16 +155,26 @@ public int compare(PatientSimilarityView o1, PatientSimilarityView o2)
      * @param referencePatient the reference patient
      * @return a query populated with terms from the patient phenotype
      */
+    @SuppressWarnings("unchecked")
     private SolrQuery generateQuery(Patient referencePatient, boolean prototypes)
     {
+        SolrQuery queryt = new SolrQuery();
+        StringBuilder qr = new StringBuilder();
+        qr.append("(document:" + ClientUtils.escapeQueryChars(referencePatient.getDocument().toString()) + " )");
```
Comment (buske) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: This crashes with an NPE when the referencePatient does not have a document (e.g., if it's a RemoteMatchingPatient). Should compute the features and ancestors on the fly.

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java [hunk 3]
```diff
@@ -151,16 +155,28 @@ public int compare(PatientSimilarityView o1, PatientSimilarityView o2)
...
     private SolrQuery generateQuery(Patient referencePatient, boolean prototypes)
     {
+        SolrQuery queryt = new SolrQuery();
+        StringBuilder qr = new StringBuilder();
+        if (referencePatient.getDocument() != null) {
+            qr.append("(document:" + ClientUtils.escapeQueryChars(referencePatient.getDocument().toString()) + " )");
+        }
+        queryt.add(CommonParams.Q, qr.toString());
+        SolrDocument doc = search(queryt).get(0);
+        List<String> ancestorsList =  (List<String>) doc.getFieldValue("phenotype_ancestors");
```
Comment (buske) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: This fails when the patient is an in-memory (e.g. remote) patient. You cannot rely on the index to get the phenotype_ancestors.

File: similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java (hunk 1 above)
Comment (allasm) [inline similarity-search/src/main/java/org/phenotips/similarity/internal/SolrSimilarPatientsFinder.java]: all other places seem to have a colon?

**Pull request closed without merging**